internal/controller/atlasschema_controller.go

```diff
@@ -187,21 +188,144 @@ func (r *AtlasSchemaReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+				delete(vars, "lint_destructive")
+			}
+			if r := p.Lint.Review; r != "" {
+				vars["lint_review"] = r
+			}
+		}
+		params := &atlasexec.SchemaApplyParams{
+			Env:    data.EnvName,
+			Vars:   vars,
+			To:     desiredURL,
+			TxMode: string(data.TxMode),
+		}
+		if repo := data.repoURL(); repo != "" {
+			// List the schema plans to check if there are any plans.
+			switch plans, err := cli.SchemaPlanList(ctx, &atlasexec.SchemaPlanListParams{
+				Env:  data.EnvName,
+				Vars: vars,
+				Repo: repo,
+				From: []string{"env://url"},
+				To:   []string{desiredURL},
+			}); {
+			case err != nil:
+				res.SetNotReady("ListingPlans", err.Error())
+				r.recorder.Event(res, corev1.EventTypeWarning, "ListingPlans", err.Error())
+				if !isSQLErr(err) {
+					err = transient(err)
+				}
+				r.recordErrEvent(res, err)
+				return result(err)
+			// There are multiple pending plans.
+			// This is an unexpected state.
+			case len(plans) > 1:
+				planURLs := make([]string, 0, len(plans))
+				for _, p := range plans {
+					planURLs = append(planURLs, p.URL)
+				}
+				log.Info("multiple schema plans found", "plans", planURLs)
+				reason, msg := "MultipleSchemaPlans", "multiple schema plans found"
+				res.SetNotReady(reason, msg)
+				r.recorder.Event(res, corev1.EventTypeWarning, reason, msg)
+				err = errors.New(msg)
+				r.recordErrEvent(res, err)
+				return result(err)
+			// Deploy the changes using the approved plan.
+			case len(plans) == 1 && plans[0].Status == "APPROVED":
+				log.Info("found an approved schema plan, applying", "plan", plans[0].URL)
+				params.PlanURL = plans[0].URL
+			// The plan is pending approval, show the plan to the user.
+			case len(plans) == 1 && plans[0].Status == "PENDING":
+				log.Info("found a pending schema plan, waiting for approval", "plan", plans[0].URL)
+				res.Status.PlanURL = plans[0].URL
+				res.Status.PlanLink = plans[0].Link
+				res.SetNotReady("ApprovalPending", "Deployment is waiting for approval")
+				r.recorder.Event(res, corev1.EventTypeNormal, "WaitingForApproval", "waiting for approval")
+				return ctrl.Result{RequeueAfter: time.Second * 5}, nil
+			}
+		}
+		// Try to apply the schema changes with lint policies,
+		// if the changes are rejected by the review policy, create a plan
+		// for the pending changes.
+		report, err = cli.SchemaApply(ctx, params)
+		// TODO: Better error handling for rejected changes.
+		if err != nil && strings.HasPrefix(err.Error(), "Rejected by review policy") {
+			log.Info("schema changes are rejected by the review policy, creating a new schema plan")
+			// The desired state is a file.
+			if data.Desired.Scheme == dbv1alpha1.SchemaTypeFile {
+				log.Info("schema is a file, pushing the schema to Atlas Cloud")
+				// We need to push the schema to the Atlas Cloud.
+				// This is to ensure that the schema is in sync with the Atlas Cloud.
+				temp, err := cli.SchemaPush(ctx, &atlasexec.SchemaPushParams{
```

Comment:
Need to verify this doesn't accidentally cause a situation where an operator loops and generates tons of junk on the server.

Reply:
genius